Автор материала: Зраев Артем.

Можно использовать в каких угодно целях.

<b> В задании нужно загрузить датасет с данными оттока и ответить на несколько вопросов (написать код). При этом сам датасет уже есть и его необязательно качать с репозитория</b>

Цель задания: проверить базовые навыки работы студентов с Pandas, умение проводить такой же базовый EDA (exploratory data analysis), делать feature engineering и обучать и валидировать модель.

https://www.kaggle.com/blastchar/telco-customer-churn

Список столбцов с типами данных в датасете:

- customerID           object
- gender               object
- SeniorCitizen         int64
- Partner              object
- Dependents           object
- tenure                int64
- PhoneService         object
- MultipleLines        object
- InternetService      object
- OnlineSecurity       object
- OnlineBackup         object
- DeviceProtection     object
- TechSupport          object
- StreamingTV          object
- StreamingMovies      object
- Contract             object
- PaperlessBilling     object
- PaymentMethod        object
- MonthlyCharges      float64
- TotalCharges         object
- Churn                object

In [77]:
import pandas as pd
import numpy as np

df = pd.read_csv("./WA_Fn-UseC_-Telco-Customer-Churn.csv")
df.head(10).T

0             1               2  \
customerID              7590-VHVEG    5575-GNVDE      3668-QPYBK   
gender                      Female          Male            Male   
SeniorCitizen                    0             0               0   
Partner                        Yes            No              No   
Dependents                      No            No              No   
tenure                           1            34               2   
PhoneService                    No           Yes             Yes   
MultipleLines     No phone service            No              No   
InternetService                DSL           DSL             DSL   
OnlineSecurity                  No           Yes             Yes   
OnlineBackup                   Yes            No             Yes   
DeviceProtection                No           Yes              No   
TechSupport                     No            No              No   
StreamingTV                     No            No              No   
StreamingMovies                 No            No              No   
Contract            Month-to-month      One year  Month-to-month   
PaperlessBilling               Yes            No             Yes   
PaymentMethod     Electronic check  Mailed check    Mailed check   
MonthlyCharges               29.85         56.95           53.85   
TotalCharges                 29.85        1889.5          108.15   
Churn                           No            No             Yes   

                                          3                 4  \
customerID                       7795-CFOCW        9237-HQITU   
gender                                 Male            Female   
SeniorCitizen                             0                 0   
Partner                                  No                No   
Dependents                               No                No   
tenure                                   45                 2   
PhoneService                             No               Yes   
MultipleLines              No phone service                No   
InternetService                         DSL       Fiber optic   
OnlineSecurity                          Yes                No   
OnlineBackup                             No                No   
DeviceProtection                        Yes                No   
TechSupport                             Yes                No   
StreamingTV                              No                No   
StreamingMovies                          No                No   
Contract                           One year    Month-to-month   
PaperlessBilling                         No               Yes   
PaymentMethod     Bank transfer (automatic)  Electronic check   
MonthlyCharges                         42.3              70.7   
TotalCharges                        1840.75            151.65   
Churn                                    No               Yes   

                                 5                        6                 7  \
customerID              9305-CDSKC               1452-KIOVK        6713-OKOMC   
gender                      Female                     Male            Female   
SeniorCitizen                    0                        0                 0   
Partner                         No                       No                No   
Dependents                      No                      Yes                No   
tenure                           8                       22                10   
PhoneService                   Yes                      Yes                No   
MultipleLines                  Yes                      Yes  No phone service   
InternetService        Fiber optic              Fiber optic               DSL   
OnlineSecurity                  No                       No               Yes   
OnlineBackup                    No                      Yes                No   
DeviceProtection               Yes                       No                No   
TechSupport                     No                  

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [30]:
df.isna().sum()


customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

> Пропуски отсутствуют

##### 1. Какое соотношение мужчин и женщин в представленном наборе данных?

In [16]:
df.gender.value_counts()

Male      3555
Female    3488
Name: gender, dtype: int64

In [24]:
df.gender[df.gender == 'Male'].count() / df.gender.count()

0.504756495811444

In [25]:
df.gender[df.gender == 'Female'].count() / df.gender.count()

0.495243504188556

> Соотношение в наборе: мужчин - 50.5%, женщин - 49.5%

##### 2. Какое количество уникальных значений у поля InternetService?

In [31]:
df.InternetService.value_counts()

Fiber optic    3096
DSL            2421
No             1526
Name: InternetService, dtype: int64

##### 3. Выведите статистики по полю TotalCharges (median, mean, std).

> TotalCharges строковое, его надо преобразовать в число

In [48]:
df.TotalCharges[df.TotalCharges.values == ''].count()

11

> Однако есть пробелы

In [ ]:
df.TotalCharges[df.TotalCharges.values == ' '] = ''

> Примем их за 0

In [49]:
df.TotalCharges[df.TotalCharges.values == ''] = 0

C:\Users\8523~1\AppData\Local\Temp/ipykernel_14324/626073563.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.TotalCharges[df.TotalCharges.values == '']=0


> Само преобразование.

In [50]:
df.TotalCharges = pd.to_numeric(df.TotalCharges, downcast="float")

In [51]:
df.TotalCharges.describe()

count    7043.000000
mean     2279.734375
std      2266.794434
min         0.000000
25%       398.549988
50%      1394.550049
75%      3786.599976
max      8684.799805
Name: TotalCharges, dtype: float64

> Статистики следующие median=1394.550049 , mean=2279.734375, std=2266.794434

В чем странность того, что вы получили? (подсказка: смотреть нужно на тип данных)

##### 4. Сделайте замену значений поля PhoneService  на числовые (Yes->1, No->0)

In [ ]:
df.PhoneService.value_counts()

In [ ]:
df.loc[df.PhoneService == 'Yes', 'PhoneService'] = 1

In [54]:
df.loc[df.PhoneService == 'No', 'PhoneService'] = 0

In [55]:
df.PhoneService.value_counts()

1    6361
0     682
Name: PhoneService, dtype: int64

##### 5. Сделайте замену пробелов в поле TotalCharges на np.nan и приведите поле к типу данных float32. Затем заполните оставшиеся пропуски значением 0 с помощью метода fillna у столбца. Снова выведите статистики и сравните с тем, что вы видели в вопросе 3

> Сделал выше

##### 6. Сделайте замену значений поля Churn на числовые (Yes -> 1, No - 0)

In [56]:
df.Churn.value_counts()


No     5174
Yes    1869
Name: Churn, dtype: int64

In [57]:
df.loc[df.Churn == 'Yes', 'Churn'] = 1

In [58]:
df.loc[df.Churn == 'No', 'Churn'] = 0


In [59]:
df.Churn.value_counts()

0    5174
1    1869
Name: Churn, dtype: int64

##### 7. Сделайте замену значений полей StreamingMovies, StreamingTV, TechSupport  на числовые (Yes -> 1, No -> 0, No internet service->0)

In [78]:
df.StreamingTV.value_counts()

No                     2810
Yes                    2707
No internet service    1526
Name: StreamingTV, dtype: int64

In [79]:
df.StreamingMovies.value_counts()

No                     2785
Yes                    2732
No internet service    1526
Name: StreamingMovies, dtype: int64

In [80]:
df.TechSupport.value_counts()

No                     3473
Yes                    2044
No internet service    1526
Name: TechSupport, dtype: int64

In [81]:
def set_val_num(df: pd.DataFrame, field, change: dict):
    for ch in change:
        df.loc[df[field] == ch] = change[ch]
    return df

In [82]:
change_val = {'Yes': 1, 'No': 0, 'No internet service': 0}
df = set_val_num(df, 'StreamingMovies', change_val)
df = set_val_num(df, 'StreamingTV', change_val)
df = set_val_num(df, 'TechSupport', change_val)

In [84]:
df.StreamingTV.value_counts()

0    4311
1    2732
Name: StreamingTV, dtype: int64

In [85]:
df.StreamingMovies.value_counts()

0    4311
1    2732
Name: StreamingMovies, dtype: int64

In [86]:
df.TechSupport.value_counts()

0    4311
1    2732
Name: TechSupport, dtype: int64

##### 8. Заполните пропуски в поле PhoneService значением 0

In [91]:
df.loc[df.PhoneService == '']



Empty DataFrame
Columns: [customerID, gender, SeniorCitizen, Partner, Dependents, tenure, PhoneService, MultipleLines, InternetService, OnlineSecurity, OnlineBackup, DeviceProtection, TechSupport, StreamingTV, StreamingMovies, Contract, PaperlessBilling, PaymentMethod, MonthlyCharges, TotalCharges, Churn]
Index: []

[0 rows x 21 columns]

#### 8. Для нашего датасета оставьте только указанный ниже список полей, удалив все другие и выведите верхние 3 строки

In [ ]:
columns = ['gender', 'tenure', 'PhoneService', 'TotalCharges',
           'StreamingMovies', 'StreamingTV', 'TechSupport', 'Churn']
#Ваш код здесь



##### 9. Разделите датасет на тренировочную и тестовую выборку (подсказка - воспользуйтесь train_test_split из sklearn.model_selection. Ссылка - https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html)

In [ ]:
from sklearn.model_selection import train_test_split

features = ['gender', 'tenure', 'PhoneService', 'TotalCharges', 'StreamingMovies', 'StreamingTV', 'TechSupport']
target = 'Churn'
#Ваш код здесь


##### 10. соберите pipeline для поля gender (нужно разобраться и изучить https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html ) из классов ColumnSelector и OHEEncoder, которые уже написаны ниже заранее

In [ ]:
from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.pipeline import Pipeline


class ColumnSelector(BaseEstimator, TransformerMixin):
    """
    Transformer to select a single column from the data frame to perform additional transformations on
    """

    def __init__(self, key):
        self.key = key

    def fit(self, X, y=None):
        return self

    def transform(self, X):
        return X[self.key]


class NumberSelector(BaseEstimator, TransformerMixin):
    """
    Transformer to select a single column from the data frame to perform additional transformations on
    Use on numeric columns in the data
    """

    def __init__(self, key):
        self.key = key

    def fit(self, X, y=None):
        return self

    def transform(self, X):
        return X[[self.key]]


class OHEEncoder(BaseEstimator, TransformerMixin):
    def __init__(self, key):
        self.key = key
        self.columns = []

    def fit(self, X, y=None):
        self.columns = [col for col in pd.get_dummies(X, prefix=self.key).columns]
        return self

    def transform(self, X):
        X = pd.get_dummies(X, prefix=self.key)
        test_columns = [col for col in X.columns]
        for col_ in test_columns:
            if col_ not in self.columns:
                X[col_] = 0
        return X[self.columns]


gender = Pipeline([
    ('selector', ColumnSelector(key='gender')),
    ('ohe', OHEEncoder(key='gender'))
])

##### 11. Вызовите метод fit_transform у пайплайна gender и передайте туда нашу тренировочную выборку (пример по ссылке из документации https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html#sklearn.pipeline.Pipeline.fit)

In [ ]:
#Ваш код здесь


##### 12. Здесь код писать уже не нужно (все сделано за вас). К полю tenure применяем StandardScaler (нормируем и центрируем). Ссылка - https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html
Вопрос - в каких случаях это может быть полезно?

In [ ]:
from sklearn.preprocessing import StandardScaler

tenure = Pipeline([
    ('selector', NumberSelector(key='tenure')),
    ('standard', StandardScaler())
])

##### 13. Напишите аналогичный (как для tenure) преобразователь поля TotalCharges

In [ ]:
#Ваш код здесь


Объединение всех "кубиков" очень легко сделать таким образом

In [ ]:
from sklearn.pipeline import FeatureUnion

number_features = Pipeline([
    ('selector', ColumnSelector(key=['PhoneService',
                                     'StreamingMovies', 'StreamingTV',
                                     'TechSupport']))
])

In [ ]:
feats = FeatureUnion([('tenure', tenure),
                      ('TotalCharges', TotalCharges),
                      ('continuos_features', number_features),
                      ('gender', gender)])
feature_processing = Pipeline([('feats', feats)])

На этом этапе что мы сделали:
1. написали преобразователь поля gender, который делает OHE кодирование
2. написали преобразователь для поля tenure, который нормирует и центрирует его 
3. повторили п. 2 для поля TotalCharges
3. для всех остальных просто взяли признаки как они есть, без изменений

У нас уже готов наш пайплайн, который преобразовывает признаки. Давайте обучим модель поверх него. В качестве модели возьмем RandomForestClassifier

In [ ]:
from sklearn.ensemble import RandomForestClassifier

pipeline = Pipeline([
    ('features', feats),
    ('classifier', RandomForestClassifier(random_state=42)),
])

pipeline.fit(X_train, y_train)

##### 14. Сделайте прогноз вероятности оттока для X_test с помощью нашего предобученного на предыдущем шаге пайплайна и убедитесь что вам возвращаются вероятности для 2 классов

In [ ]:
#Ваш код здесь


##### 15. Посчитайте метрики качества получившейся модели (roc_auc, logloss)

In [ ]:
from sklearn.metrics import roc_auc_score, log_loss

#Ваш код здесь


### Сохраним наш пайплайн

In [ ]:
import dill

with open("model_RF.dill", "wb") as f:
    dill.dump(pipeline, f)